src/skydb.ts

```diff
@@ -1,105 +1,79 @@
+  // TODO: Replace with download request method.
+  let skylink: string;
+  try {
+    skylink = parseSkylink(entry.value.data);
+  } catch (error) {
+    throw new Error(`invalid skylink: ${error}`);
+  }
+
   const response = await this.executeRequest({
     ...this.customOptions,
     method: "get",
     url: this.getSkylinkUrl(skylink),
   });
 
-  // wrap the data in a skyfile and return it
-  const metadata = JSON.parse(response.headers["skynet-file-metadata"]);
-  const file = new SkyFile(new File([response.data], metadata.filename, { type: response.headers["content-type"] }));
-  return file;
+  return { json: response.data, revision: entry.revision };
 }
 
-// setFile uploads a file and sets updates the registry
-export async function setFile(this: SkynetClient, user: User, fileID: FileID, file: SkyFile): Promise<boolean> {
-  // upload the file to acquire its skylink
-  const customFilename = fileID.filename;
-  const skylink = await this.uploadFile(file.file, { customFilename });
-
-  // fetch the current value to find out the revision, timeout after 2s
-  let existing: SignedRegistryValue | null;
-  try {
-    existing = await promiseTimeout(this.lookupRegistry(user, fileID), 2000);
-  } catch (error) {
-    existing = null;
+export async function setJSON(this: SkynetClient, privateKey: SecretKey, dataKey: string, json: object, revision: number = -1): Promise<boolean> {
+  // Upload the data to acquire its skylink
+  // TODO: Replace with upload request method.
+  const response = await this.executeRequest({
+    ...this.customOptions,
+    method: "post",
+    endpointPath: "/skynet/skyfile",
+    data: json.toString(),
+  });
+  const skylink = response.skylink;
+
+  const publicKey = pki.ed25519.publicKeyFromPrivateKey({privateKey});
+  if (revision === -1) {
+    // fetch the current value to find out the revision.
+    const entry = await this.registry.lookup(publicKey, dataKey)
+
+    if (entry) {
+      // verify here
+      if (!pki.ed25519.verify({
+        message: HashRegistryEntry(entry.entry),
+        signature: entry.signature,
+        publicKey,
+      })) {
+        throw new Error("could not verify signature");
+      }
+
+      revision = entry.revision + 1;
+    } else {
+      revision = 0;
+    }
   }
 
-  // TODO: we could (/should?) verify here
-
   // build the registry value
-  const value: RegistryValue = {
-    tweak: HashFileID(fileID),
+  const entry: RegistryEntry = {
     data: trimUriPrefix(skylink, uriSkynetPrefix),
-    revision: existing ? existing.value.revision + 1 : 0,
+    revision,
   };
 
   // sign it
-  const signature = user.sign({ message: HashRegistryValue(value) });
+  const signature = pki.ed25519.sign({
+    message: HashRegistryEntry(entry),
+    privateKey,
+  });
 
   // update the registry
-  const updated = await this.updateRegistry(user, fileID, { value, signature });
+  const updated = await this.registry.update(publicKey, dataKey, entry, signature);
   return updated;
```

Comment:
No need to await returned value.

```suggestion
  return this.registry.update(publicKey, dataKey, entry, signature);
```